### ISSUE TITLE
[feature request] Publish wheels with debug symbols

### ISSUE THREAD

**Issue by nicoulaj** [pytorch/pytorch#11982]
It would be great if wheels of pytorch with debug symbols were published (or at least made available via https://ci.pytorch.org), so that users can investigate native crashes without going through the whole (error prone) process of building pytorch themselves.

For reference, I asked [the question on the forums first, but got no answer](https://discuss.pytorch.org/t/pytorch-wheel-with-debug-symbols/25169).

**Comment by zou3519** [comment]
cc @pjh5 what do you think?

**Comment by nicoulaj** [comment]
Linking to #6391 as I think it is related

**Comment by pjh5** [comment]
I built a set of wheels with debug symbols once for some internal debugging. IIRC it added about 8 hours on each of our nightly machines and generated another 40 GB. Given the wide array of possible build configurations and that debug symbols are probably most useful when built with your particular environment, I think that building from source for debug symbols is actually the better option here.

**Comment by zou3519** [comment]
Closing via @pjh5's comments

**Comment by zou3519** [comment]
@pjh5 What is the difference between a wheel and a build from source? I remember there was some distinction (static linking of cuda bits?)

**Comment by ezyang** [comment]
> debug symbols are probably most useful when built with your particular environment

This is not true. Debug symbols are always useful, even if you're running a binary on another environment.

SOP in Linux distro land is to distribute debug symbols separately from executable binaries, since they are pretty big. We might have trouble setting infrastructure for this. Maybe if we can make it easy for end users to build wheels, we can make this self service, which will be easier for us.

Reopening.

**Comment by pjh5** [comment]
Right now the wheels are built with the pytorch/builder repo, which contains Dockerfiles and bash scripts that call setup.py and then do some linking and path munging. It comes down to setting a few environment variables in a Docker https://github.com/pytorch/builder/blob/master/cron/build_docker.sh#L142 (DESIRED_PYTHON and DESIRED_CUDA and DEBUG are all that should matter) and then calling https://github.com/pytorch/builder/blob/master/manywheel/build.sh (or build_cpu.sh). You should be able to produce a debug wheel with this in the exact same conditions, except perhaps for differences in nvidia-docker versions or CUDA drivers.

**Comment by nicoulaj** [comment]
@pjh5 Thanks, I could easily build with debug symbols using:
```
export PACKAGE_TYPE=manywheel
export DESIRED_PYTHON=3.6m
export DESIRED_CUDA=cu90
export PYTORCH_BUILD_VERSION=0.4.1
export DEBUG=1
export NIGHTLIES_EMAIL_LIST=''
export NIGHTLIES_FOLDER=$(mktemp -d)
./cron/build_docker.sh
```
The scripts are not really intented to do this (notifications and such) but it  works. As @ezyang said, maybe all is needed is to make these build scripts a bit more accessible or documented.

**Comment by pjh5** [comment]
cron/build_docker.sh is quite coupled to the nightlies setup. But https://github.com/pytorch/builder/blob/master/manywheel/build.sh will build a wheel without needing to set most of those environment variables if you run it in a soumith/manylinux-cuda80 (you'll have to checkout the pytorch/builder repo but it comes with git). I think you'd only have to set DESIRED_PYTHON and DESIRED_CUDA. That way it won't require a nightlies folder to store all the nightlies metadata and logs

**Comment by ezyang** [comment]
This issue is still valid but IDK if anyone is actually going to get around to doing it.

**Comment by malfet** [comment]
Actually, we do plan to do something soon in that area.
cc: @driazati

**Comment by keepdying** [comment]
wow, its 2024 and still no wheel with debug symbols 💀

**Comment by abhillman** [comment]
This would be still be really helpful for debugging issues using gdb/lldb

**Comment by abhillman** [comment]
```sh
[nix-shell:~]$ cat ok.py 
import torch
torch.manual_seed(1337)
print(torch.randint(1000, (1,)))

[nix-shell:~]$ lldb python3
(lldb) target create "python3"
Current executable set to 'python3' (x86_64).
(lldb) b native::randint
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run ok.py
Process 194123 launched: '/nix/store/8k4p193rnzy3rqjkv19abz3b738l25z9-python3-3.11.6/bin/python3' (x86_64)
4 locations added to breakpoint 1
Process 194123 stopped
* thread #1, name = 'python3', stop reason = breakpoint 1.4
    frame #0: 0x00007fffec3717c0 libtorch_cpu.so`at::native::randint(long, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>)
libtorch_cpu.so`at::native::randint:
->  0x7fffec3717c0 <+0>: pushq  %rbp
    0x7fffec3717c1 <+1>: movq   %rsp, %rbp
    0x7fffec3717c4 <+4>: pushq  %rbx
    0x7fffec3717c5 <+5>: movq   %rdi, %rbx
(lldb) 
```

**Comment by wenjianhn** [comment]
I think symbols are helpful for debugging this DDP OOM issue: https://github.com/pytorch/pytorch/issues/106294

1. disable linux memory cgroup OOM for pytorch containers
2. gdb time

**Comment by abhillman** [comment]
I think one of the bigger questions is how to get debug symbols at all. Even with a local build using `DEBUG=1`, symbols do not appear to be retained atm against the latest release with gcc11.
